types/index.d.ts

```diff
@@ -0,0 +1,136 @@
+import inquirer = require('inquirer');
+// @types/globby doesn't export types for GlobOptions, so we have to work a little bit to extract them:
+// GlobOptions is the second parameter of the sync function, which can be extracted with the Parameters<T> type
+import { sync as _sync } from 'globby';
+type GlobOptions = Parameters<typeof _sync>[1];
+import { HelperDelegate as HelperFunction } from 'handlebars';
+
+interface NodePlopAPI {
+	getGenerator(name: string): PlopGenerator;
+	setGenerator(name: string, config: PlopGenerator): PlopGenerator;
+
+	setPrompt(name: string, prompt: inquirer.PromptModule): void;
+	setWelcomeMessage(message: string): void;
+	getWelcomeMessage(): string;
+	getGeneratorList(): { name: string; description: string }[];
+	setPartial(name: string, str: string): void;
+	getPartial(name: string): string;
+	getPartialList(): string[];
+	setHelper(name: string, fn: HelperFunction): void;
+	getHelper(name: string): Function;
+	getHelperList(): string[];
+	setActionType(name: string, fn: CustomActionFunction): void;
+	getActionType(name: string): ActionType;
+	getActionTypeList(): string[];
+
+	setPlopfilePath(filePath: string): void;
+	getPlopfilePath(): string;
+	getDestBasePath(): string;
+
+	// plop.load functionality
+	load(target: string[] | string, loadCfg: PlopCfg, includeOverride: boolean): void;
+	setDefaultInclude(inc: object): void;
+	getDefaultInclude(): object;
+
+	renderString(template: string, data: any): string; //set to any matching handlebars declaration
+
+	// passthroughs for backward compatibility
+	addPrompt(name: string, prompt: inquirer.PromptModule): void;
+	addPartial(name: string, str: string): void;
+	addHelper(name: string, fn: Function): void;
+}
+
+interface PlopActionHooksFailures {
+	type: string;
+	path: string;
+	error: string;
+	message: string;
+}
+
+interface PlopActionHooksChanges {
+	type: string;
+	path: string;
+}
+
+interface PlopActionHooks {
+	onComment?: (msg: string) => void;
+	onSuccess?: (change: PlopActionHooksChanges) => void;
+	onFailure?: (failure: PlopActionHooksFailures) => void;
+}
+
+export interface PlopGenerator {
+	description: string;
+	prompts: inquirer.Question[];
+	actions: ActionType[];
```

Comment:
we need to make sure Actions and Prompts accepts their Dynamic versions, as I also believe a lot of projects depends on those.

Reply:
I completely agree but wanted to leave that out-of-scope for this PR. This isn't meant to solve too many typing issues, simply provide a baseline testing utility so that we're able to fix the typings further in a way that's verifiable (simply by dropping-in valid code into the testing file)

Reply:
@crutchcorn got it! sounds like a good reasoning indeed. `plop` isn't going to be affected anyway as long as you don't bump the version of `node-plop` there, so I think it is safe too